bolt-modules/out/lib/puppet/functions/out/message.rb

```diff
@@ -24,9 +26,12 @@ def output_message(message)
         .from_issue_and_stack(Bolt::PAL::Issues::PLAN_OPERATION_NOT_SUPPORTED_WHEN_COMPILING, action: 'out::message')
     end
 
+    stringified_message = Bolt::Util::Format.stringify(message)
+
     Puppet.lookup(:bolt_executor).tap do |executor|
       executor.report_function_call(self.class.name)
-      executor.publish_event(type: :message, message: Bolt::Util::Format.stringify(message))
+      executor.publish_event(type: :message, message: stringified_message)
+      executor.publish_event(type: :log, level: :info, message: stringified_message)
```

Comment:
I don't think this is strictly necessary, but it would be nice (and doable) if this didn't print twice if the log level was such that the logger logs to the console. I think the best way to accomplish that is to have a helper method in util that inspects `Logging.logger[:root].appenders.select { |a| a.name == 'console' }.first.level` to get the level of the console log (in theory there could be a logger for stdout and stderr but that doesn't ever happen in Bolt), and use that here to see if `Bolt::Util.console_log_level =< Logging::level_num(:info)` and *not* log in that case (the levels go down for more verbosity, so `info = 2`, `warn = 3`, `error = 4`, etc.)

If that doesn't ✨ just work ✨ I don't think we have to worry about it.

Reply:
That would prevent the message from printing in log files as well, which I don't think we want. We could probably skip the `:message`/`:verbose` events, though I'm not 100% sure if that might mess things up anywhere. It seems like it should work, since the logger prints to stderr with the json format...

Fwiw, Bolt already prints some messages to the console twice depending on the log level:

```shell
$be bolt plan run bolt --log-level info

Loaded default project from '/Users/tom/.puppetlabs/bolt'
Starting: plan bolt
Starting: plan bolt
hello
hello
Finished: plan bolt in 0.0 sec
Finished: plan bolt in 0.0 sec
Plan completed successfully with no result
```

I'm fine with suppressing the `:message`/`:verbose` events if you are.

Reply:
No no, I meant to only print the log event and not the message event, not not do the log event. That check should also happen in the Outputter instead of here, so we always send both events and the outputter knows to not print a message twice if it's already been printed.

I know it's not a big deal, but like I said seems doable and makes sense for users.

Reply:
I guess the last thing we should be aware of is that this means `out::*` will print to _either_ stdout or stderr, depending on the configured log level.

Reply:
Just messing around with the layout for the console logs, I wonder if adding the log level before log messages might help?

```console
INFO: Loaded default project from '/Users/tom/.puppetlabs/bolt'
WARN: CLI arguments ["host-key-check"] might be overridden by Inventory: /Users/tom/.puppetlabs/bolt/inventory.yaml [ID: cli_overrides]
Starting: plan bolt
INFO: Starting: plan bolt
hello
INFO: hello
Finished: plan bolt in 0.01 sec
INFO: Finished: plan bolt in 0.01 sec
Plan completed successfully with no result
```

Just throwing out ideas at this point. :P

Reply:
That's definitely not a bad idea... https://github.com/beechtom/bolt/pull/22 this is along the lines of what I'm thinking. The only potential issue I see is that we don't know if stdout or stderr are getting redirected or used in a meaningful way, so inspecting both and then "merging" could mean that we don't log something to stdout based on the level of stderr, even if stderr is getting sent somewhere else, so the message isn't seen. I think a better but also riskier tact would be to close and reopen the stderr logger around the message event if it has type `:message`, but Bolt is multithreaded and I'm not confident that won't have it's own issues. I think we can leave it as is for now and if people complain we can go back to the drawing board.

Reply:
I know what you meant and did a [similar implementation](https://github.com/puppetlabs/bolt/commit/e3e4724562d3610d50c64390c289760a56786c40) last week. :) After writing that I realized the weirdness with inconsistent printing between stdout/stderr.

I would definitely prefer to leave it printing twice for now and see if users would prefer the duplicate message to be suppressed.